src/unix/pty.cc

```diff
@@ -131,6 +132,131 @@ pty_after_waitpid(uv_async_t *);
 static void
 pty_after_close(uv_handle_t *);
 
+int pty_fork_classic(char **argv, char **env, char *cwd, int uid, int gid, const termios *term, winsize *winp, int &pty_master, pid_t &pid) {
+  sigset_t newmask, oldmask;
+  struct sigaction sig_action;
+
+  // temporarily block all signals
+  // this is needed due to a race condition in openpty
+  // and to avoid running signal handlers in the child
+  // before exec* happened
+  sigfillset(&newmask);
+  pthread_sigmask(SIG_SETMASK, &newmask, &oldmask);
+
+  pid = pty_forkpty(&pty_master, nullptr, term, winp);
+
+  if (!pid) {
+    // remove all signal handler from child
+    sig_action.sa_handler = SIG_DFL;
+    sig_action.sa_flags = 0;
+    sigemptyset(&sig_action.sa_mask);
+    for (int i = 0 ; i < NSIG ; i++) {    // NSIG is a macro for all signals + 1
+      sigaction(i, &sig_action, NULL);
+    }
+  }
+  // reenable signals
+  pthread_sigmask(SIG_SETMASK, &oldmask, NULL);
+
+  if (pid) {
+    if (pid == -1) {
+      Nan::ThrowError("forkpty(3) failed.");
+      return 1;
+    }
+
+    return 0;
+  } else {
+    if (strlen(cwd)) {
+      if (chdir(cwd) == -1) {
+        perror("chdir(2) failed.");
+        _exit(1);
+      }
+    }
+
+    if (uid != -1 && gid != -1) {
+      if (setgid(gid) == -1) {
+        perror("setgid(2) failed.");
+        _exit(1);
+      }
+      if (setuid(uid) == -1) {
+        perror("setuid(2) failed.");
+        _exit(1);
+      }
+    }
+
+    pty_execvpe(argv[0], argv, env);
+
+    perror("execvp(3) failed.");
+    _exit(1);
+  }
+  return 0;
+}
+
+int pty_fork_posix_spawn(char **argv, char **env, const termios *term, winsize *winp, int &pty_master, pid_t &pid) {
+  sigset_t all_signals, oldmask;
+
+  // temporarily block all signals
+  // this is needed due to a race condition in openpty
+  // and to avoid running signal handlers in the child
+  // before exec* happened
+  sigfillset(&all_signals);
+  pthread_sigmask(SIG_SETMASK, &all_signals, &oldmask);
+
+  int pty_slave;
+  if (pty_openpty(&pty_master, &pty_slave, nullptr, term, winp) == -1) {
+    pthread_sigmask(SIG_SETMASK, &oldmask, NULL);
+    Nan::ThrowError("pty_openpty failed.");
+    return 1;
+  }
+
+  pthread_sigmask(SIG_SETMASK, &oldmask, NULL);
+
+  posix_spawn_file_actions_t acts;
+  posix_spawnattr_t attrs;
+  if (posix_spawn_file_actions_init(&acts)) {
+    Nan::ThrowError("posix_spawn_file_actions_init failed.");
+    return 1;
+  }
+  if (posix_spawnattr_init(&attrs)) {
+    Nan::ThrowError("posix_spawnattr_init failed.");
+    return 1;
+  }
+
+  if (
+    posix_spawn_file_actions_adddup2(&acts, pty_slave, STDIN_FILENO) ||
+    posix_spawn_file_actions_adddup2(&acts, pty_slave, STDOUT_FILENO) ||
+    posix_spawn_file_actions_addclose(&acts, pty_slave) ||
+    posix_spawn_file_actions_addclose(&acts, pty_master)
```

Comment:
I think you missed STDERR here, should also point to pty_slave.

Furthermore nodejs holds several fds open, I think those should be marked for closing as well. Its a bit cumbersome to find them, basically needs looping up to RLMIIT_NOFILE and checking if its a valid fd with `fcntl(fd, F_GETFD)`.

Closing all other fds is currently not done for the normal fork way either, thus you can also just ignore my remark to stay inline with the fork way. Imho this is a security nightmare, esp. when coming from a higher privileged user, but the issue in the past about this led to nowhere.

Reply:
I've just tried this, and `posix_spawn_file_actions_addclose` mysteriously fails on macOS when doing this. There's also no way to debug since posix_spawn will fail whenever there's _any_ error executing the file actions...

One way could be to set `FD_CLOEXEC` on every descriptor before spawning and reset it later, but that's prone to races

Reply:
What I tried was
```c
  struct rlimit rlim_ofile;
  getrlimit(RLIMIT_NOFILE, &rlim_ofile);
  for (int fd = STDERR_FILENO + 1; fd < rlim_ofile.rlim_cur; fd++) {
    if (fcntl(fd, F_GETFD) == -1) {
      if (posix_spawn_file_actions_addclose(&acts, fd)) {
        posix_spawn_file_actions_destroy(&acts);
        Nan::ThrowError("posix_spawn_file_actions_addclose failed.");
        return 1;
      }
    }
  }
```

Reply:
Oh, macOS has `POSIX_SPAWN_CLOEXEC_DEFAULT`

Reply:
`(fcntl(fd, F_GETFD) == -1)` --> `(fcntl(fd, F_GETFD) != -1)`? Otherwise this looks good to me.

Reply:
Done - `POSIX_SPAWN_CLOEXEC_DEFAULT` is set when available, otherwise FDs are left unclosed, same as in the fork path